// @pixi/layers/lib/index.js
/*!
 * @pixi/layers - v2.0.1
 * Compiled Wed, 21 Dec 2022 21:17:52 UTC
 *
 * @pixi/layers is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * 
 * Copyright 2017-2021, Ivan Popelyshev, All Rights Reserved
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var DisplayMixin = require('./DisplayMixin.js');
var RendererMixin = require('./RendererMixin.js');
var core = require('@pixi/core');
var Stage = require('./Stage.js');
var Layer = require('./Layer.js');
var Group = require('./Group.js');

DisplayMixin.applyDisplayMixin();
RendererMixin.applyRendererMixin(core.Renderer);

exports.applyContainerRenderMixin = DisplayMixin.applyContainerRenderMixin;
exports.applyDisplayMixin = DisplayMixin.applyDisplayMixin;
exports.applyParticleMixin = DisplayMixin.applyParticleMixin;
exports.applyCanvasMixin = RendererMixin.applyCanvasMixin;
exports.applyRendererMixin = RendererMixin.applyRendererMixin;
exports.Stage = Stage.Stage;
exports.Layer = Layer.Layer;
exports.LayerTextureCache = Layer.LayerTextureCache;
exports.Group = Group.Group;
