// @pixi/lights/lib/lights/light/LightShader.js
/*!
 * @pixi/lights - v4.0.0
 * Compiled Thu, 22 Dec 2022 00:00:35 UTC
 *
 * @pixi/lights is licensed under the MIT License.
 * http://www.opensource.org/licenses/mit-license
 * 
 * Copyright 2017-2021, Ivan Popelyshev, All Rights Reserved
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@pixi/core');
var mesh = require('@pixi/mesh');
var shared = require('../shared.js');

class LightShader extends mesh.MeshMaterial {
  constructor(options) {
    const uniforms = {
      translationMatrix: core.Matrix.IDENTITY.toArray(true),
      uNormalSampler: core.Texture.WHITE,
      uViewSize: new Float32Array(2),
      uViewPixels: new Float32Array(2),
      uLightFalloff: new Float32Array([0, 0, 0]),
      uLightHeight: 0.075,
      uBrightness: 1,
      uUseViewportQuad: true
    };
    Object.assign(uniforms, options?.uniforms);
    super(core.Texture.WHITE, { ...options, uniforms });
  }
}
LightShader.defaultVertexSrc = shared.vert;

exports.LightShader = LightShader;
